### ISSUE TITLE
Add beforeLoad lifecycle callback

### ISSUE THREAD

**Issue by goldami1** [spring-projects/spring-data-mongodb#3891]
Hi,

Following #3334 , I suggest to add additionally to the afterSave, afterConvert and beforeSave callbacks, also the beforeLoad.
Similarly to how AfterLoadEvent is emited before the reader is used to read the document (in ReadDocumentCallback).
Thus, Similar functionality would be apprepciated at the Spring Data EntityCallbacks level for both ES and MONGO.

(for example can be named like: maybeCallAfterLoad(document, type, collectionName))

![image](https://user-images.githubusercontent.com/33942741/143085833-74c2d6eb-6ae1-4528-95e1-309c67a10353.png)
![image](https://user-images.githubusercontent.com/33942741/143085845-dc50e9d3-d8ff-4753-9170-f5729c555894.png)

**Comment by mp911de** [comment]
Care to elaborate? MongoDB supports `AfterLoadEvent`/`AfterConvertEvent` and many more. A `BeforeLoad` without a document doesn't make much sense.

**Comment by goldami1** [comment]
@mp911de why `BeforeLoad` without a document? line 3177 is publishing `AfterLoadEvent` with a concrete document before reader is triggered for further domain object translation.
Thus, I suggest to add a maybeCallAfterLoad with the same document as published in `AfterLoadEvent` right after the event publishing.
(at this moment, document is definitely not null according to line 3176)

**Comment by mp911de** [comment]
`AfterLoadEvent` is called after getting hold of a `Document` so `BeforeLoad` would need to happen before that, right? Likely there's already some confusion here.

In any case, having an entity callback for `AfterLoad` isn't possible because we do not have an entity instance at that time and entity callbacks (as you can tell from the name) are driven by entities.

**Comment by goldami1** [comment]
I think the confusion is caused due to callback naming. I meant `AfterLoad` callback which would be parallel to `AfterLoadEvent`.
However, I do understand that due to this callback being an Entity callback, it supposes to hold an entity. which makes it unreasonable to be used without an entity.
On the other hand, this callback triggering prior to domain object creation can be useful.

For example my scenario for this matter is as following:
I have mongodb&elastic dbs which hold an entity containing multiple enum fields. every one of these enums has different format. I would like to standardize all of these prior the reader starts to deserialize them.
I though about implementing an Enum ConverterFactory which will handle it and register this factory to the springData's conversionService.
However this will couple the enums themselves with this deserialization. while I would prefer to decouple it, and make the deserialization binding on the field level. (achieve similar functionality as with the jackson's @JsonDeserialize)

**Comment by mp911de** [comment]
Thanks for the additional context.

For the time being, if you use the synchronous event multicaster, then you're good to post-process the `Document` in `AfterLoadEvent`. Alternatively, you can wrap or subclass `MappingMongoConverter` and add your customizations there.

**Comment by goldami1** [comment]
@mp911de Does it make sense this mapping lifecycle part to be part of the mapping callbacks? (some similar callbacks mechanism to EntityCallBack mechanism)

**Comment by caspian-borison-by** [comment]
@mp911de 
I realize this is perhaps better suited as a feature request (and it makes the assumption that #3902 is a valid issue as well).

However, reading this, I realized there might be a viable use case for a beforeLoad event/callback which works in this way:
* executed only when a specific ID is being used (e.g. findById)
* returns a Document object that, if non null, bypasses the actual mongo find() query

Such a callback, coupled with an appropriate "onAfterLoad" event (to manage the entity reference cache) could be used to allow applications to create a short lived first-level caching mechanism to avoid repeated reads of entities when loading them via @DocumentReference annotations.

**Comment by mp911de** [comment]
> Such a callback, coupled with an appropriate "onAfterLoad" event (to manage the entity reference cache) could be used to allow applications to create a short lived first-level caching mechanism to avoid repeated reads of entities when loading them via @DocumentReference annotations.

We do not have hooks to intercept such lookups, that would be rather an aspect that needs to be implemented in the reference resolver as interception hooks are not part of the what this was designed for. It's also not possible to achieve this in an easy way to drop in a callback and be done with it.

This ticket isn't developing into an actionable direction, so we're closing it for now.
